### DISCUSSION TITLE
v0.1.0-pre1

### DISCUSSION THREAD

**Discussion by mrene** [mrene/minidsp-rs#110] [General]
This is a pre-release of the next version.

#### `minidspd` daemon replaces `minidsp server` with a configurable service
When using the debian package, a default configuration is installed at `/etc/minidsp/config.toml`. This configuration only listens on 127.0.0.1 and can be edited to expose the api or the plugin-compatible server to the local network.

The daemon will auto-discover WI-DGs on the local network, no further configuration is necessary to use these.

#### An HTTP API is now available
The default configuration automatically listens on 127.0.0.1:5380, and on /tmp/minidsp.sock (for unix platforms). See `README-API.md` for a summary of available APIs.

####  On unix platforms, the `minidsp` command line app detects the locally running service and automatically uses it 
No more errors because the USB device is already in used by the server component.

#### New devices supported
The miniSHARC 4x8 is now auto-detected and supports the full feature set.
Other devices supported in [this list](https://www.minidsp.com/products/accessories/wi-dg) will work but only allow changing the master status (config preset, source, master volume & mute)

<hr /><em>This discussion was created from the release <a href='https://github.com/mrene/minidsp-rs/releases/tag/v0.1.0-pre1'>v0.1.0-pre1</a>.</em>

**Comment by 3ll3d00d** [comment]
looks great

a couple of thoughts

* I think a ws api for updates would be v useful 
* the POSTs feel like PATCHs to me 
* the readme-api (or schema calls) don't comment on return status codes, might be useful to add something here?
* are updates to device config atomic? put another way, if I specify n updates then does it process them 1 by 1 and blow up at first error or would it validate the request 1st then apply or does it rollback on error?

**Reply by mrene** [reply]
There will be a proper openapi docs for all of these, with schemas for both success and error responses. Although there isn't so much that can fail, short of the device hanging or being disconnected in the middle of an update.

Good call on `PATCH`, there's probably something idiomatic to use here.

Updates are applied sequentially, the entire request is validated at first and the whole operation bailing out at the first (hardware) error. I haven't seen any device error feedback outside of the "this command opcode isn't recognized". Most parameters are blindly writing stuff to the dsp's memory  - if values are out of range the device behaviour is undefined.

It's also going to let you do Bad Things(tm) if you send concurrent update requests that mutate the device state in conflicting ways, there will need to be some locking at some point but it gets tedious with the plugin app capable of being connected.

Rolling anything back isn't possible until enough data is kept to know what the previous state was.

**Reply by mrene** [reply]
This doc should be a little easier to read https://minidsp-rs.mrene.workers.dev/

**Reply by mrene** [reply]
Websocket is implemented, streaming changes done through the IR remote (due to some protocol differences, this won't show updates done through the plugin app, or through minidsp-rs)
```
$ websocat ws://127.0.0.1:5380/devices/0                                                                                                                                   
{"master":{"preset":0,"source":"Toslink","volume":-19.5,"mute":false},"input_levels":[-44.400307,-42.08899],"output_levels":[-54.93488,-56.41222,-120.0,-120.0]}
{"master":{"volume":-19.0,"mute":false}}
{"master":{"volume":-18.5,"mute":false}}
{"master":{"volume":-18.0,"mute":false}}
```

**Reply by kukocz** [reply]
Hi,
With the websocat 1.8.0 I got the following response:

```
~ $ websocat ws://127.0.0.1:5380/devices/0
websocat: WebSocketError: Received unexpected status code (200 OK)
websocat: error running
```
Do you have any idea what is it related to? Tried with two other versions of websocat with no luck.

I also noticed another issue:

> On unix platforms, the minidsp command line app detects the locally running service and automatically uses it

When service is started with connected minidsp, then I've got error: `Error: No device found`
Reboot doesn't hel. The only way to fix it is to reconnect minidsp (while service is running). The workaround for that is to use `minidsp --tcp "ip_addr"` like in previous versions.

**Reply by mrene** [reply]
These were implemented after 0.1.0-pre1, try this build https://github.com/mrene/minidsp-rs/releases/tag/v0.1.0-pre2

The issue of using minidsp while the service is running is also fixed (it didn't have world read/write permissions).

**Comment by kukocz** [comment]
Hi mrene. You've done really great job!
I'm using minidsp-rs with raspberry pi zero to control volume in my DDRC-24 and display it on OLED display.
To allow display volume and other parameters on the display (including buttons/rotary encoder control as well as remote control) I have to read status periodically what could be done max ~10 times per second. I noticed it is really CPU consuming.

What is interesting, when using minidsp server mode and connecting native desktop app - volume is being updated continuously with no effect on CPU on minidsp server side.

Is there any way to set up server mode with continuous reading (at 5-10 times per second) which could be read from the pipeline and processed by another script (i.e. display control)?

As I'm using DDRC-24 with Dirac Live - it would be great to have possibility to control DL status (on/off) via command line.

Best Regards!

**Reply by mrene** [reply]
Hey! That looks like an interesting setup. 

Are you just polling for the current master volume (and source, preset, mute)? These get asynchronous device events, so it's possible to get them without actively polling the device.

For current meter levels (for inputs, outputs, compressors), these can be polled to display the amount of signal going through the device (that's what the app is polling a few times a second, when you have the associated section of the plugin opened)

I was planning on adding these as websocket streams, so a webapp could connect, indicate what its interested in, and get async events. I could put it in the command line tool as well, which you would call as something like `minidsp -o jsonline events --hz 3 --inputs --outputs --compressors` and poll them a few times a second. What are you using to drive the display part?

Feel free to open an issue for dirac support, I've seen some forum threads that have the address so it should be quick to add. 

One more thing, could you send me the output of `minidsp debug id`? Especially the top part.. I'll make it so it recognizes the DDRC-24 correctly :) 

It'll look like this:
```
HW ID: 10
DSP Version: 100
Detected sources: [(Analog, 0), (Toslink, 1), (Usb, 2)]
```

**Reply by kukocz** [reply]
Hi,
I'm sorry for late asnwer.

> Are you just polling for the current master volume (and source, preset, mute)? These get asynchronous device events, so it's possible to get them without actively polling the device.

Yes, very basic python script to run separate thread and send simple "minidsp" command few times per second to retrieve the current status. It's cpu-consuming :)

> For current meter levels (for inputs, outputs, compressors), these can be polled to display the amount of signal going through the device (that's what the app is polling a few times a second, when you have the associated section of the plugin opened)
>
> I was planning on adding these as websocket streams, so a webapp could connect, indicate what its interested in, and get async events. I could put it in the command line tool as well, which you would call as something like minidsp -o jsonline events --hz 3 --inputs --outputs --compressors and poll them a few times a second. 

That sounds brilliant! Looking forward for such functionality in your further versions

>What are you using to drive the display part?

I was inspired with this project: [MiniDSP Preamp](https://gitlab.com/andrejg/minidsp-preamp) but Andrej used separate microcontroller to drive the OLED display. Also some values have been stored locally without minidsp feedback (only read once on startup). 
I wanted to get rid of additional device and was looking around for another solution to control display directly via Pi Zero.
I have the [OLED display 128x64 ](https://www.diymore.cc/products/2-42-inch-12864-oled-display-module-iic-i2c-spi-serial-for-arduino-c51-stm32-green-white-blue-yellow?variant=17060396433466) based on SSD1309 driver.
Finally I was able to make it working with [luma.oled](https://github.com/rm-hull/luma.oled) and [py-gaugette](https://github.com/guyc/py-gaugette) for rotary encoder control. 
I'm planning to publish my code on github some day, but it is still in pre-alpha stage. Moreover I'm rather an ethusiast than professional programmer with very limited knowledge of python , therefore my code wouldn't be optimized for sure :)

> Feel free to open an issue for dirac support, I've seen some forum threads that have the address so it should be quick to add.

I'll do that :)

> One more thing, could you send me the output of minidsp debug id? Especially the top part.. I'll make it so it recognizes the DDRC-24 correctly :)

Here's the response:
```
Probing local hid devices: 
Found: [2752:0044] path=0001:0004:04

HW ID: 10
DSP Version: 101
Detected sources: [(Analog, 0), (Toslink, 1), (Usb, 2)]

Dumping memory:
len=95
0000FFA0  FF 65 FF 03 20 03 2C 01 03 00 64 00 00 00 00 00  | .e.♥ ♥,☺♥.d..... |
0000FFB0  00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  | ................ |
0000FFC0  00 00 FF FF F8 FF FF FF 05 DC A6 77 04 F0 D4 B6  | ....▫...♣▄ªw♦≡╘╢ |
0000FFD0  01 81 5F 42 02 10 3A 3B 01 01 4E 01 FF FF FF FF  | ☺ü_B☻►:;☺☺N☺.... |
0000FFE0  00 00 00 00 00 00 FF FF FF FF FF FF FF FF FF FF  | ................ |
0000FFF0  FF FF FF FF FF FF FF FF FF FF FF FF FF FF 06     | ..............♠  |
```
Maybe this link could be also useful (info about command/response for minidsp settings)
[Volume display - minidsp forum](https://www.minidsp.com/forum/suggestion-box/14442-volume-display#38356)

![minidsp_OLED](https://user-images.githubusercontent.com/82149485/117768006-03fdb680-b232-11eb-8af7-04b516ea5ba3.jpg)

**Reply by liselotte111** [reply]
@kukocz I want to achieve exactly the same thing! Would you be willing to share?

**Comment by 3ll3d00d** [comment]
the example response in in https://minidsp-rs.pages.dev/daemon/http#1-list-available-devices includes `"url": "tcp://192.168.1.242:5333?name=Living%20Room%20Pi"`

where does that name come from? i don't see it mentioned anywhere else

context here being thinking about how it will expose multiple connected devices (and whether there is some reliable name or id I can refer to)

**Reply by mrene** [reply]
The URL is transport-dependent. If it's via USB it'll represent the device path in an OS-specific way. The information includes the device's serial number, which should, when not garbled and reset to 65535 by random bugs, be somewhat reliable.

**Comment by 3ll3d00d** [comment]
is it expected to be possible to use the websocket over the unix socket or just the http based api?

**Reply by mrene** [reply]
the Unix socket is the exact same http server as the tcp one, if you detect a file present and can connect to it, it's easier to connect via Unix.

This is the default command line behaviour too, it makes it easier to work with widg devices because they only send advertisement packets so often.

**Reply by mrene** [reply]
This crude example taken from websocket-client's docs work:

```python3
import socket
from websocket import create_connection

def main():
    my_socket = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    my_socket.connect("/tmp/minidsp.sock")

    ws = create_connection("ws://localhost/devices/0?levels=true", socket = my_socket)
    while True:
        try:
            msg = ws.recv()
            print(msg)
        except Exception as e:
            print(e)

if __name__ == "__main__":
    main()
```

Prints out:
```json
{"master":{"preset":0,"source":"Toslink","volume":-34.0,"mute":false},"input_levels":[-131.49413,-131.49413],"output_levels":[-166.56247,-166.24278,-120.0,-120.0]}
{"master":{},"input_levels":[-131.47736,-131.47736],"output_levels":[-166.51753,-166.20457,-120.0,-120.0]}
{"master":{},"input_levels":[-131.4203,-131.4203],"output_levels":[-166.42278,-166.11952,-120.0,-120.0]}
{"master":{},"input_levels":[-131.40881,-131.40881],"output_levels":[-166.39104,-166.09303,-120.0,-120.0]}
```

**Reply by 3ll3d00d** [reply]
ok thanks, I raised https://github.com/mrene/minidsp-rs/issues/134 to cover the issue I was seeing as it seems more like some bug to me

**Comment by liselotte111** [comment]
Since it wasn't a success to run the http server from windows, I've switched to Linux. It all works fine (after building inside my volumio RPI4), however after some time (it varies between 30 min and several hours) the http connection times out:

`minidspd[25624]: [2021-08-17T08:08:47Z ERROR hyper::server::tcp] accept error: Too many open files (os error 24)
`
I'm polling the status every 30s using RestFul in Home Assistant.

System info:
Linux volumio 4.19.118-v7l+ #1311 SMP Mon Apr 27 14:26:42 BST 2020 armv7l GNU/Linux on RPI 4

Edit:
now polling via nodered, no more issues
